sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -342,11 +342,12 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
    */
   private def extractCorrelatedScalarSubqueries[E <: Expression](
       expression: E,
-      subqueries: ArrayBuffer[ScalarSubquery]): E = {
+      subqueries: ArrayBuffer[(ScalarSubquery, ExprId)]): E = {
     val newExpression = expression transform {
       case s: ScalarSubquery if s.children.nonEmpty =>
-        subqueries += s
-        s.plan.output.head
+        val newExprId = NamedExpression.newExprId
+        subqueries += s -> newExprId
+        s.plan.output.head.withExprId(newExprId)
```

Comment:
we need more comments to explain what's going on here: we will rewrite the subqueries later and reference their results here by a new attribute.

Reply:
BTW I feel it's more natural to replace attributes after the subquery rewriting, instead of generating the expr id here.

Reply:
> BTW I feel it's more natural to replace attributes after the subquery rewriting, instead of generating the expr id here.

Yea, that refactoring looks reasonable and the current logic (top-down rewriting) looks a bit weird. But, I wanna keep the current logic in this PR in order to avoid making new bugs. Is it okay to do it in a separate PR?

Reply:
Added comments there.

Reply:
If it's top-down now, let's defer to change it and fix it in a new PR.